src/ipc/client.cpp

```diff
@@ -0,0 +1,175 @@
+
+Handler::~Handler() {}
+
+void Handler::disconnect() const
+{
+    auto request = impl->handlerClient.disconnectRequest();
+    auto promise = request.send();
+    promise.wait(impl->ioContext.waitScope);
+}
+
+//! Wallet private member struct.
+struct Wallet::Impl {
+    kj::AsyncIoContext& ioContext;
+    messages::Wallet::Client walletClient;
+    Impl(kj::AsyncIoContext& ioContext_, messages::Wallet::Client walletClient_) : ioContext(ioContext_), walletClient(std::move(walletClient_)) {}
+};
+
+Wallet::Wallet(std::unique_ptr<Impl> impl_) : impl(std::move(impl_)) {}
+
+Wallet::~Wallet() {}
+
+CAmount Wallet::getBalance() const
+{
+    auto request = impl->walletClient.getBalanceRequest();
+    auto promise = request.send();
+    auto response = promise.wait(impl->ioContext.waitScope);
+    return response.getValue();
+}
+
+//! Node private member struct.
+struct Node::Impl {
+    kj::AsyncIoContext ioContext{kj::setupAsyncIo()};
+    kj::Own<kj::AsyncIoStream> clientStream;
+    capnp::TwoPartyVatNetwork clientNetwork{*clientStream, capnp::rpc::twoparty::Side::CLIENT, capnp::ReaderOptions()};
+    capnp::RpcSystem<capnp::rpc::twoparty::VatId> rpcClient{capnp::makeRpcClient(clientNetwork)};
+    messages::Node::Client nodeClient{rpcClient.bootstrap(ServerVatId().vatId).castAs<messages::Node>()};
+
+    Impl(int fd) : clientStream(ioContext.lowLevelProvider->wrapSocketFd(fd, kj::LowLevelAsyncIoProvider::TAKE_OWNERSHIP)) {}
+};
+
+Node::Node(std::unique_ptr<Impl> impl_) : impl(std::move(impl_)) {}
+
+Node::~Node() {}
+
+std::string Node::helpMessage(HelpMessageMode mode) const
+{
+    auto request = impl->nodeClient.helpMessageRequest();
+    request.setMode(mode);
+    auto promise = request.send();
+    auto response = promise.wait(impl->ioContext.waitScope);
+    return response.getValue();
+}
+
+std::unique_ptr<Handler> Node::handleInitMessage(std::function<void(const std::string&)> callback) const
+{
+    auto request = impl->nodeClient.handleInitMessageRequest();
+    request.setCallback(kj::heap<InitMessageCallbackServer>(std::move(callback)));
+    auto promise = request.send();
+    auto response = promise.wait(impl->ioContext.waitScope);
+    return Factory::MakeImpl<Handler>(impl->ioContext, response.getHandler());
+}
+
+std::unique_ptr<Wallet> Node::wallet() const
+{
+    auto request = impl->nodeClient.walletRequest();
+    auto promise = request.send();
+    auto response = promise.wait(impl->ioContext.waitScope);
+    return Factory::MakeImpl<Wallet>(impl->ioContext, response.getWallet());
+}
+
+std::unique_ptr<Node> StartClient()
+{
+    int fds[2];
+    KJ_SYSCALL(socketpair(AF_UNIX, SOCK_STREAM, 0, fds));
+
+    if (fork() == 0) {
+        int maxFd = MaxFd();
+        for (int fd = 3; fd < maxFd; ++fd) {
+            if (fd != fds[0]) {
+                close(fd);
+            }
+        }
+        if (execlp(BITCOIN_DAEMON_NAME, BITCOIN_DAEMON_NAME, "-ipcfd", std::to_string(fds[0]).c_str(), nullptr) != 0) {
```

Comment:
Right, ideally bitcoin-qt would need only a very small subset of the bitcoind code.